libs/providers/multi-provider/src/lib/multi-provider.ts

```diff
@@ -0,0 +1,393 @@
+
+// Represents an entry in the constructor's provider array which may or may not have a name set
+export type ProviderEntryInput = {
+  provider: Provider;
+  name?: string;
+};
+
+// Represents a processed and "registered" provider entry where a name has been chosen
+type RegisteredProvider = {
+  name: string;
+  provider: Provider;
+};
+
+export class MultiProvider implements Provider {
+  readonly runsOn = 'server';
+
+  public readonly events = new OpenFeatureEventEmitter();
+
+  private hookContexts: WeakMap<EvaluationContext, HookContext> = new WeakMap<EvaluationContext, HookContext>();
+  private hookHints: WeakMap<EvaluationContext, HookHints> = new WeakMap<EvaluationContext, HookHints>();
+
+  metadata: ProviderMetadata;
+
+  providerEntries: RegisteredProvider[] = [];
+  private providerEntriesByName: Record<string, RegisteredProvider> = {};
+  private readonly providerStatuses: Record<string, ProviderStatus> = {};
+
+  private hookExecutor: HookExecutor;
+
+  constructor(
+    readonly constructorProviders: ProviderEntryInput[],
+    private readonly evaluationStrategy: BaseEvaluationStrategy = new FirstMatchStrategy(),
+    private readonly logger: Logger = new DefaultLogger(),
+  ) {
+    this.hookExecutor = new HookExecutor(this.logger);
+
+    this.registerProviders(constructorProviders);
+
+    // just make sure we don't accidentally modify these later
+    Object.freeze(this.providerEntries);
+    Object.freeze(this.providerEntriesByName);
+
+    this.providerStatuses = {};
+
+    const aggregateMetadata = Object.keys(this.providerEntriesByName).reduce((acc, name) => {
+      return { ...acc, [name]: this.providerEntriesByName[name].provider.metadata };
+    }, {});
+
+    this.metadata = {
+      ...aggregateMetadata,
+      name: MultiProvider.name,
+    };
+  }
+
+  private registerProviders(constructorProviders: ProviderEntryInput[]) {
+    const providersByName: Record<string, Provider[]> = {};
+
+    for (const constructorProvider of constructorProviders) {
+      const providerName = constructorProvider.provider.metadata.name;
+      let candidateName = constructorProvider.name ?? providerName;
+
+      if (constructorProvider.name && providersByName[constructorProvider.name]) {
+        throw new Error('Provider names must be unique');
+      }
+
+      providersByName[candidateName] ||= [];
+      providersByName[candidateName].push(constructorProvider.provider);
+    }
+
+    for (const name of Object.keys(providersByName)) {
+      const useIndexedNames = providersByName[name].length > 1;
+      for (let i = 0; i < providersByName[name].length; i++) {
+        const indexedName = useIndexedNames ? `${name}-${i + 1}` : name;
+        this.providerEntriesByName[indexedName] = { provider: providersByName[name][i], name: indexedName };
+        this.providerEntries.push(this.providerEntriesByName[indexedName]);
+        this.wrapEventHandler(this.providerEntriesByName[indexedName]);
```

Comment:
why not just pull out `providersByName[name][i]` into: 
```
const provider = providersByName[name][i]
```
would make the rest of the code easier to read.